File: src/dotnet/Ansight.Tools.Reflection/InspectObjectTool.cs
namespace Ansight.Tools.Reflection;

public sealed class InspectObjectTool : ITool
{
    private readonly ReflectionToolsOptions options;

    public InspectObjectTool(ReflectionToolsOptions? options = null)
    {
        this.options = options ?? ReflectionToolsOptions.Default;
    }

    public string Category => "reflect";

    public ToolScope Scope => ToolScope.Read;

    public string Id => ReflectionToolIds.InspectObject;

    public string Name => "Inspect Object";

    public string Description => "Inspects a registered live object and returns a stateless expandable snapshot.";

    public string Keywords => "reflection inspect object runtime properties fields methods";

    public ToolSchema ArgumentsSchema => ReflectionToolSchemas.InspectObjectArguments;

    public ToolSchema ResultSchema => ReflectionToolSchemas.InspectObjectResult;

    public ToolSecurity Security => ReflectionToolSecurityProfiles.InspectObject;

    public Task<ToolResult> Execute(IReadOnlyDictionary<string, string> arguments)
    {
        ArgumentNullException.ThrowIfNull(arguments);

        try
        {
            return Task.FromResult(ToolResult.Success(ReflectionSupport.InspectObject(options, arguments)));
        }
        catch (Exception exception)
        {
            return Task.FromResult(ToolResult.Failure(exception.Message, errorCode: "reflect_inspect_failed"));
        }
    }
}

File: src/ios/Sources/AnsightToolsReflection/InspectObjectTool.swift
import AnsightCore
import Foundation

public final class InspectObjectTool: AnsightTool {
    private let runtime: AnsightRuntime
    private let options: AnsightReflectionToolsOptions

    public init(
        options: AnsightReflectionToolsOptions = .default,
        runtime: AnsightRuntime = .shared
    ) {
        self.runtime = runtime
        self.options = options
    }

    public var descriptor: AnsightToolDescriptor {
        AnsightToolDescriptor(
            id: AnsightReflectionToolIds.inspectObject,
            name: "Inspect Object",
            description: "Inspects a registered iOS object root and returns an expandable snapshot.",
            category: "reflect",
            scope: AnsightToolScope.read.rawValue,
            keywords: "reflection inspect object runtime properties fields methods",
            security: AnsightReflectionToolSecurityProfiles.inspectObject,
            argumentsSchema: AnsightReflectionToolSchemas.inspectObjectArguments,
            resultSchema: AnsightReflectionToolSchemas.inspectObjectResult
        )
    }

    public func execute(arguments: [String: String]) throws -> AnsightToolExecutionResult {
        do {
            return .success(try AnsightReflectionSupport.inspectObject(runtime: runtime, options: options, arguments: arguments))
        } catch {
            return .failure(error.localizedDescription, errorCode: "reflect_inspect_failed")
        }
    }
}
